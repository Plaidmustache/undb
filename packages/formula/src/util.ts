import { CharStreams, CommonTokenStream } from "antlr4ts"
import { FormulaLexer } from "./grammar/FormulaLexer"
import { FormulaParser } from "./grammar/FormulaParser"
import { CustomFormulaVisitor } from "./visitor"

export function createParser(input: string) {
  const inputStream = CharStreams.fromString(input)
  const lexer = new FormulaLexer(inputStream)
  const tokenStream = new CommonTokenStream(lexer)
  return new FormulaParser(tokenStream)
}

export function parseFormula(input: string) {
  const parser = createParser(input)

  const tree = parser.formula()

  const visitor = new CustomFormulaVisitor()
  const parsedFormula = visitor.visit(tree)

  return parsedFormula
}
