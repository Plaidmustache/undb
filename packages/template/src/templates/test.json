{
  "name": "hello",
  "tables": {
    "world": {
      "schema": {
        "name": {
          "type": "string",
          "constraint": {
            "max": 10
          },
          "defaultValue": "world",
          "display": true
        }
      }
    }
  }
}
