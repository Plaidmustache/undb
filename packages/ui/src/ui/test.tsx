export const Test = () => {
  return <div>{new Date().toISOString()}</div>
}
