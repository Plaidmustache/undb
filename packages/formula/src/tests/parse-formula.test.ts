import { describe, expect, test } from "bun:test"
import { parseFormula } from "../util"

describe("parse formula", () => {
  test.each([
    //
    "ADD(1, ADD(2, {{ field1 }}))",
    "ADD(1, 2)",
    "SUBTRACT(1, 2)",
    "MULTIPLY(1, 2)",
    "DIVIDE(1, 2)",
    "1 - 1",
    "1 * 1",
    "1 / 1",
    "SUBTRACT(1, 2) + MULTIPLY(3, 4)",
    "1",
    "{{field1}}",
    "1 + 1",
    "{{field1}} + {{field2}}",
    "SUM({{field1}}, {{field2}})",
    "CONCAT({{field1}}, {{field2}})",
    "CONCAT({{field1}}, {{field2}}, {{field3}})",
  ])("test %s", (input) => {
    const result = parseFormula(input)

    expect(result).toMatchSnapshot()
  })
})
