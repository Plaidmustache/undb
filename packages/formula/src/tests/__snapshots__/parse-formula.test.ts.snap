// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parse formula test ADD(1, ADD(2, {{ field1 }})) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1,
    },
    {
      "arguments": [
        {
          "type": "number",
          "value": 2,
        },
        {
          "type": "variable",
          "value": "{{field1}}",
          "variable": "field1",
        },
      ],
      "name": "ADD",
      "returnType": "number",
      "type": "functionCall",
      "value": "ADD(2,{{field1}})",
    },
  ],
  "name": "ADD",
  "returnType": "number",
  "type": "functionCall",
  "value": "ADD(1,ADD(2,{{field1}}))",
}
`;

exports[`parse formula test ADD(1, 2) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1,
    },
    {
      "type": "number",
      "value": 2,
    },
  ],
  "name": "ADD",
  "returnType": "number",
  "type": "functionCall",
  "value": "ADD(1,2)",
}
`;

exports[`parse formula test SUBTRACT(1, 2) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1,
    },
    {
      "type": "number",
      "value": 2,
    },
  ],
  "name": "SUBTRACT",
  "returnType": "number",
  "type": "functionCall",
  "value": "SUBTRACT(1,2)",
}
`;

exports[`parse formula test MULTIPLY(1, 2) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1,
    },
    {
      "type": "number",
      "value": 2,
    },
  ],
  "name": "MULTIPLY",
  "returnType": "number",
  "type": "functionCall",
  "value": "MULTIPLY(1,2)",
}
`;

exports[`parse formula test DIVIDE(1, 2) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1,
    },
    {
      "type": "number",
      "value": 2,
    },
  ],
  "name": "DIVIDE",
  "returnType": "number",
  "type": "functionCall",
  "value": "DIVIDE(1,2)",
}
`;

exports[`parse formula test 1 - 1 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1,
    },
    {
      "type": "number",
      "value": 1,
    },
  ],
  "name": "-",
  "returnType": "number",
  "type": "functionCall",
  "value": "1-1",
}
`;

exports[`parse formula test 1 * 1 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1,
    },
    {
      "type": "number",
      "value": 1,
    },
  ],
  "name": "*",
  "returnType": "number",
  "type": "functionCall",
  "value": "1*1",
}
`;

exports[`parse formula test 1 / 1 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1,
    },
    {
      "type": "number",
      "value": 1,
    },
  ],
  "name": "/",
  "returnType": "number",
  "type": "functionCall",
  "value": "1/1",
}
`;

exports[`parse formula test SUBTRACT(1, 2) + MULTIPLY(3, 4) 1`] = `
{
  "arguments": [
    {
      "arguments": [
        {
          "type": "number",
          "value": 1,
        },
        {
          "type": "number",
          "value": 2,
        },
      ],
      "name": "SUBTRACT",
      "returnType": "number",
      "type": "functionCall",
      "value": "SUBTRACT(1,2)",
    },
    {
      "arguments": [
        {
          "type": "number",
          "value": 3,
        },
        {
          "type": "number",
          "value": 4,
        },
      ],
      "name": "MULTIPLY",
      "returnType": "number",
      "type": "functionCall",
      "value": "MULTIPLY(3,4)",
    },
  ],
  "name": "+",
  "returnType": "number",
  "type": "functionCall",
  "value": "SUBTRACT(1,2)+MULTIPLY(3,4)",
}
`;

exports[`parse formula test 1 1`] = `
{
  "type": "number",
  "value": 1,
}
`;

exports[`parse formula test {{field1}} 1`] = `
{
  "type": "variable",
  "value": "{{field1}}",
  "variable": "field1",
}
`;

exports[`parse formula test 1 + 1 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1,
    },
    {
      "type": "number",
      "value": 1,
    },
  ],
  "name": "+",
  "returnType": "number",
  "type": "functionCall",
  "value": "1+1",
}
`;

exports[`parse formula test {{field1}} + {{field2}} 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
    {
      "type": "variable",
      "value": "{{field2}}",
      "variable": "field2",
    },
  ],
  "name": "+",
  "returnType": "number",
  "type": "functionCall",
  "value": "{{field1}}+{{field2}}",
}
`;

exports[`parse formula test SUM({{field1}}, {{field2}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
    {
      "type": "variable",
      "value": "{{field2}}",
      "variable": "field2",
    },
  ],
  "name": "SUM",
  "returnType": "number",
  "type": "functionCall",
  "value": "SUM({{field1}},{{field2}})",
}
`;

exports[`parse formula test CONCAT({{field1}}, {{field2}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
    {
      "type": "variable",
      "value": "{{field2}}",
      "variable": "field2",
    },
  ],
  "name": "CONCAT",
  "returnType": "string",
  "type": "functionCall",
  "value": "CONCAT({{field1}},{{field2}})",
}
`;

exports[`parse formula test CONCAT({{field1}}, {{field2}}, {{field3}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
    {
      "type": "variable",
      "value": "{{field2}}",
      "variable": "field2",
    },
    {
      "type": "variable",
      "value": "{{field3}}",
      "variable": "field3",
    },
  ],
  "name": "CONCAT",
  "returnType": "string",
  "type": "functionCall",
  "value": "CONCAT({{field1}},{{field2}},{{field3}})",
}
`;

exports[`parse formula test MOD(1, 2) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1,
    },
    {
      "type": "number",
      "value": 2,
    },
  ],
  "name": "MOD",
  "returnType": "number",
  "type": "functionCall",
  "value": "MOD(1,2)",
}
`;

exports[`parse formula test MOD({{field1}}, {{field2}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
    {
      "type": "variable",
      "value": "{{field2}}",
      "variable": "field2",
    },
  ],
  "name": "MOD",
  "returnType": "number",
  "type": "functionCall",
  "value": "MOD({{field1}},{{field2}})",
}
`;

exports[`parse formula test POWER(2, 3) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 2,
    },
    {
      "type": "number",
      "value": 3,
    },
  ],
  "name": "POWER",
  "returnType": "number",
  "type": "functionCall",
  "value": "POWER(2,3)",
}
`;

exports[`parse formula test POWER({{field1}}, {{field2}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
    {
      "type": "variable",
      "value": "{{field2}}",
      "variable": "field2",
    },
  ],
  "name": "POWER",
  "returnType": "number",
  "type": "functionCall",
  "value": "POWER({{field1}},{{field2}})",
}
`;

exports[`parse formula test SQRT(4) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 4,
    },
  ],
  "name": "SQRT",
  "returnType": "number",
  "type": "functionCall",
  "value": "SQRT(4)",
}
`;

exports[`parse formula test SQRT({{field1}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
  ],
  "name": "SQRT",
  "returnType": "number",
  "type": "functionCall",
  "value": "SQRT({{field1}})",
}
`;

exports[`parse formula test ABS(-5) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 5,
    },
  ],
  "name": "ABS",
  "returnType": "number",
  "type": "functionCall",
  "value": "ABS(-5)",
}
`;

exports[`parse formula test ABS({{field1}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
  ],
  "name": "ABS",
  "returnType": "number",
  "type": "functionCall",
  "value": "ABS({{field1}})",
}
`;

exports[`parse formula test ROUND(1.234) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1.234,
    },
  ],
  "name": "ROUND",
  "returnType": "number",
  "type": "functionCall",
  "value": "ROUND(1.234)",
}
`;

exports[`parse formula test ROUND({{field1}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
  ],
  "name": "ROUND",
  "returnType": "number",
  "type": "functionCall",
  "value": "ROUND({{field1}})",
}
`;

exports[`parse formula test FLOOR(1.234) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1.234,
    },
  ],
  "name": "FLOOR",
  "returnType": "number",
  "type": "functionCall",
  "value": "FLOOR(1.234)",
}
`;

exports[`parse formula test FLOOR({{field1}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
  ],
  "name": "FLOOR",
  "returnType": "number",
  "type": "functionCall",
  "value": "FLOOR({{field1}})",
}
`;

exports[`parse formula test CEILING(1.234) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1.234,
    },
  ],
  "name": "CEILING",
  "returnType": "number",
  "type": "functionCall",
  "value": "CEILING(1.234)",
}
`;

exports[`parse formula test CEILING({{field1}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
  ],
  "name": "CEILING",
  "returnType": "number",
  "type": "functionCall",
  "value": "CEILING({{field1}})",
}
`;

exports[`parse formula test MIN(1, 2) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1,
    },
    {
      "type": "number",
      "value": 2,
    },
  ],
  "name": "MIN",
  "returnType": "number",
  "type": "functionCall",
  "value": "MIN(1,2)",
}
`;

exports[`parse formula test MIN({{field1}}, {{field2}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
    {
      "type": "variable",
      "value": "{{field2}}",
      "variable": "field2",
    },
  ],
  "name": "MIN",
  "returnType": "number",
  "type": "functionCall",
  "value": "MIN({{field1}},{{field2}})",
}
`;

exports[`parse formula test MIN({{field1}}, {{field2}}, {{field3}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
    {
      "type": "variable",
      "value": "{{field2}}",
      "variable": "field2",
    },
    {
      "type": "variable",
      "value": "{{field3}}",
      "variable": "field3",
    },
  ],
  "name": "MIN",
  "returnType": "number",
  "type": "functionCall",
  "value": "MIN({{field1}},{{field2}},{{field3}})",
}
`;

exports[`parse formula test MAX(1, 2) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1,
    },
    {
      "type": "number",
      "value": 2,
    },
  ],
  "name": "MAX",
  "returnType": "number",
  "type": "functionCall",
  "value": "MAX(1,2)",
}
`;

exports[`parse formula test MAX({{field1}}, {{field2}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
    {
      "type": "variable",
      "value": "{{field2}}",
      "variable": "field2",
    },
  ],
  "name": "MAX",
  "returnType": "number",
  "type": "functionCall",
  "value": "MAX({{field1}},{{field2}})",
}
`;

exports[`parse formula test MAX({{field1}}, {{field2}}, {{field3}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
    {
      "type": "variable",
      "value": "{{field2}}",
      "variable": "field2",
    },
    {
      "type": "variable",
      "value": "{{field3}}",
      "variable": "field3",
    },
  ],
  "name": "MAX",
  "returnType": "number",
  "type": "functionCall",
  "value": "MAX({{field1}},{{field2}},{{field3}})",
}
`;

exports[`parse formula test AVERAGE(1, 2, 3) 1`] = `
{
  "arguments": [
    {
      "type": "number",
      "value": 1,
    },
    {
      "type": "number",
      "value": 2,
    },
    {
      "type": "number",
      "value": 3,
    },
  ],
  "name": "AVERAGE",
  "returnType": "number",
  "type": "functionCall",
  "value": "AVERAGE(1,2,3)",
}
`;

exports[`parse formula test AVERAGE({{field1}}, {{field2}}, {{field3}}) 1`] = `
{
  "arguments": [
    {
      "type": "variable",
      "value": "{{field1}}",
      "variable": "field1",
    },
    {
      "type": "variable",
      "value": "{{field2}}",
      "variable": "field2",
    },
    {
      "type": "variable",
      "value": "{{field3}}",
      "variable": "field3",
    },
  ],
  "name": "AVERAGE",
  "returnType": "number",
  "type": "functionCall",
  "value": "AVERAGE({{field1}},{{field2}},{{field3}})",
}
`;
